modules/ml/src/main/java/org/apache/ignite/ml/environment/LearningEnvironmentBuilder.java

```diff
@@ -17,109 +17,135 @@
+    /**
+     * Builds {@link LearningEnvironment} for worker on given partition.
+     *
+     * @param part Partition.
+     * @return {@link LearningEnvironment} for worker on given partition.
+     */
+    default public LearningEnvironment buildForWorker(int part) {
+        return buildForTrainer();
+    }
 
     /**
-     * Creates an instance of LearningEnvironmentBuilder.
+     * Builds learning environment for trainer.
+     *
+     * @return Learning environment for trainer.
      */
-    public LearningEnvironmentBuilder() {
-        parallelismStgy = NoParallelismStrategy.INSTANCE;
-        loggingFactory = NoOpLogger.factory();
+    default public LearningEnvironment buildForTrainer() {
+        return buildForWorker(-1);
     }
 
     /**
-     * Specifies Parallelism Strategy for LearningEnvironment.
+     * Specifies dependency (partition -> Parallelism Strategy Type for LearningEnvironment).
      *
-     * @param stgy Parallelism Strategy.
+     * @param stgyType Function describing dependency (partition -> Parallelism Strategy Type).
+     * @return This object.
      */
-    public <T> LearningEnvironmentBuilder withParallelismStrategy(ParallelismStrategy stgy) {
-        this.parallelismStgy = stgy;
+    public LearningEnvironmentBuilder withParallelismStrategyType(
+        IgniteFunction<Integer, ParallelismStrategy.Type> stgyType);
 
-        return this;
+    /**
+     * Specifies Parallelism Strategy Type for LearningEnvironment. Same strategy type will be used for all partitions.
+     *
+     * @param stgyType Parallelism Strategy Type.
+     * @return This object.
+     */
+    public default LearningEnvironmentBuilder withParallelismStrategyType(ParallelismStrategy.Type stgyType) {
+        return withParallelismStrategyType(constant(stgyType));
     }
 
     /**
-     * Specifies Parallelism Strategy for LearningEnvironment.
+     * Specifies dependency (partition -> Parallelism Strategy for LearningEnvironment).
      *
-     * @param stgyType Parallelism Strategy Type.
+     * @param stgy Function describing dependency (partition -> Parallelism Strategy).
+     * @return This object.
+     */
+    public LearningEnvironmentBuilder withParallelismStrategy(IgniteFunction<Integer, ParallelismStrategy> stgy);
+
+    /**
+     * Specifies Parallelism Strategy for LearningEnvironment. Same strategy type will be used for all partitions.
+     *
+     * @param stgy Parallelism Strategy.
+     * @return This object.
+     */
+    public default LearningEnvironmentBuilder withParallelismStrategy(ParallelismStrategy stgy) {
+        return withParallelismStrategy(constant(stgy));
+    }
+
+
+    /**
+     * Specify dependency (partition -> logging factory).
+     *
+     * @param loggingFactory Function describing (partition -> logging factory).
+     * @return This object.
+     */
+    public LearningEnvironmentBuilder withLoggingFactory(IgniteFunction<Integer, MLLogger.Factory> loggingFactory);
```

Comment:
Let's suppose that logging with metrics aggregation may be huge process. If we bind factory to partition then such huge processes will be started several times on one node. I think that on this step of architecture this is not problem but we should consider this later.

Reply:
I think here we should give user most precise level of granularity, it is her choice to blur it (maybe even ignore it) in needed ways (for example `part -> Ignition.localIgnite().cluster().nodeLocalMap().computeIfAbsent(LOG_AGGREGATOR_FACTORY_FOR_DATASET, key -> someFactorySupplier)`)

Reply:
We should give user maximal level of granularity, it is user choice to blur it (in most extreme case, ignore it). With this API we still can make grouping by nodes instead of paritions.